### ISSUE TITLE
mipsel targets: compilation fails with "Undefined temporary symbol $BB0_4"

### ISSUE THREAD

**Issue by Mrmaxmeier** [rust-lang/rust#108835]
Hi,

I hit a strange build failure in CI where  `rustc-demangle` fails to build for `mipsel-unknown-linux-gnu` .

Here's a reduced reproducer:
```rust
// ⋊> rustc ./repro.rs --crate-type=lib --target mipsel-unknown-linux-gnu
// error: Undefined temporary symbol $BB0_4
//
// error: aborting due to previous error
pub fn func(s: &()) -> usize {
    #[allow(dead_code)]
    enum Error {
        VarA,
        VarB,
    }

    fn subfunc<'a>(_s: &'a ()) -> Result<&'a str, Error> {
        loop {}
    }
    match subfunc(s) {
        Ok(_d) => 1,
        Err(Error::VarA) | Err(Error::VarB) => 2,
    }
}
```

### Meta

I'm using current nightly, but this is a regression from stable (`1.67.1`) to beta (`1.68.0-beta.7`).

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (7820b62d2 2023-03-05)
binary: rustc
commit-hash: 7820b62d20bc548096d4632a3487987308cb4b5d
commit-date: 2023-03-05
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

`cargo bisect-rustc` points to d726c8467c06088d5d4488edf6b015ec9698c1ea (in `nightly-2022-11-01`), but this feels like an LLVM bug.
I can't trigger it via `--emit=llvm-ir` and `llc repro.ll -o repro.o -filetype=obj` though. The LLVM version matches, but I'm probably not reproducing `rustc`'s interactions with LLVM faithfully..

Sorry for the noise with obscure target architectures, but I'm reporting the issue here because I don't know how to reproduce it on LLVM's side :slightly_smiling_face:

**Comment by workingjubilee** [comment]
> Sorry for the noise with obscure target architectures

It's fine. We may not offer the best support for them due to technical and resource constraints, but we still want people to have the option to run Rust code on their OpenWRT router or their PlayStation or whatever.

**Comment by cnjn** [comment]
i hit the same problem. build rustc-demangle failed for mipsel-unknown-linux-gnu

**Comment by workingjubilee** [comment]
May be caused by a trigger for https://github.com/rust-lang/rust/issues/113065

**Comment by taiki-e** [comment]
FYI, this can be worked around by enabling optimization.

```console
$ cargo +1.67 build --target mipsel-unknown-linux-gnu
   Compiling repro v0.1.0 (...)
error: Undefined temporary symbol $BB0_4

error: could not compile `repro` due to previous error

$ cargo +1.67 build --target mipsel-unknown-linux-gnu --release
   Compiling repro v0.1.0 (...)
    Finished release [optimized] target(s) in 0.09s

$ RUSTFLAGS='-C opt-level=1' cargo +1.67 build --target mipsel-unknown-linux-gnu
   Compiling repro v0.1.0 (...)
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
``` 

The same workaround works for error from `rustc-demangle` when using build-std in the latest nightly. (https://github.com/taiki-e/rust-cross-toolchain/commit/e820504ebfdccf542ff619723acb410241163022, https://github.com/taiki-e/setup-cross-toolchain-action/commit/1ca8c89e498e672632f69f2fe5728a7780da427f, https://github.com/taiki-e/atomic-maybe-uninit/commit/ec71c70e6d4a223d49b06dda3f9d6bf1db7fe05d, https://github.com/taiki-e/portable-atomic/commit/d47811eb91d75dfb9705e3a1f058e1a81a36e57b)

**Comment by Fearyncess** [comment]
> FYI, this can be worked around by enabling optimization.

Maybe its relocation is too far then that has been truncated?

**Comment by urosbericsyrmia** [comment]
This issue should be fixed by [llvm/llvm-project#80799](https://github.com/llvm/llvm-project/pull/80799).
Initially, the cause of the issue and the solution are explained in [llvm/llvm-project#80666](https://github.com/llvm/llvm-project/pull/80666), but more elegant solution is just to remove derived isBlockOnlyReachableByFallthrough for Mips architecture.

**Comment by Mrmaxmeier** [comment]
Yep. My initial reproducer fails with `rustc 1.76.0 (07dca489a 2024-02-04)` but compiles on `1.78.0-nightly (a84bb95a1 2024-02-13)`. :chart_with_upwards_trend:
